### PR TITLE
Fix ClusterStateRestCancellationIT

### LINKED ISSUE #72056: [CI] ClusterStateRestCancellationIT testClusterStateRestCancellation trips assertion
It hasn't failed before in this way, but felt worth filing since it tripped an `assert` in the code:

```
    if (acceptableClusterStatePredicate.test(state)) {
            ActionListener.completeWith(listener, () -> buildResponse(request, state));
    } else {
            assert acceptableClusterStateOrFailedPredicate.test(state) == false;
    ...
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/awg667j3ynola/tests/:qa:smoke-test-http:integTest/org.elasticsearch.http.ClusterStateRestCancellationIT/testClusterStateRestCancellation

**Reproduction line:**
`./gradlew ':qa:smoke-test-http:integTest' --tests "org.elasticsearch.http.ClusterStateRestCancellationIT.testClusterStateRestCancellation" -Dtests.seed=BF5860E36BE09C8F -Dtests.locale=hr-HR -Dtests.timezone=America/Boise -Druntime.java=15`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:** this test has failed once before recently, but not with the same error

**Failure excerpt:**
```
org.elasticsearch.http.ClusterStateRestCancellationIT > classMethod FAILED
    com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=1526, name=elasticsearch[node_s0][transport_worker][T#3], state=RUNNABLE, group=TGRP-ClusterStateRestCancellationIT]

        Caused by:
        java.lang.AssertionError
            at __randomizedtesting.SeedInfo.seed([BF5860E36BE09C8F]:0)
            at org.elasticsearch.action.admin.cluster.state.TransportClusterStateAction.masterOperation(TransportClusterStateAction.java:82)
            at org.elasticsearch.action.admin.cluster.state.TransportClusterStateAction.masterOperation(TransportClusterStateAction.java:42)
            at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.lambda$doStart$3(TransportMasterNodeAction.java:151)
            at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
```

### PR DESCRIPTION
**Pull request by fcofdez**
Take into account that the task can be cancelled after the check for
acceptableClusterStateOrFailedPredicate has been performed and 
it could trip the assertion.

Closes #72056

### CODE REVIEW

**Comment by fcofdez** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java [hunk 1]
```diff
@@ -79,7 +79,10 @@ protected void masterOperation(Task task, final ClusterStateRequest request, fin
         if (acceptableClusterStatePredicate.test(state)) {
             ActionListener.completeWith(listener, () -> buildResponse(request, state));
         } else {
-            assert acceptableClusterStateOrFailedPredicate.test(state) == false;
+            // It is possible that the task is cancelled after the predicate has been executed, therefore we should take
+            // that into account as well for the assertion
+            assert acceptableClusterStateOrFailedPredicate.test(state) == false
+                || request.local() == false && cancellableTask.isCancelled();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java:85]: I was expecting the `cancellableTask.isCancelled()` here but not the `request.local()` check. Digging into why that's needed led me to conclude that there's a bug in #67413: local requests should also complete on the first cluster state update after cancellation.

Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java]: You're right, I've moved the cancellation check outside of `acceptableClusterStateOrFailedPredicate` making it deterministic given the same `ClusterState`. Now we check for cancellation in local requests too.

File: qa/smoke-test-http/src/test/java/org/elasticsearch/http/ClusterStateRestCancellationIT.java [hunk 2]
```diff
@@ -71,6 +71,9 @@ public void testClusterStateRestCancellation() throws Exception {
         final Request clusterStateRequest = new Request(HttpGet.METHOD_NAME, "/_cluster/state");
         clusterStateRequest.addParameter("wait_for_metadata_version", Long.toString(Long.MAX_VALUE));
         clusterStateRequest.addParameter("wait_for_timeout", "1h");
+        if (randomBoolean()) {
+            clusterStateRequest.addParameter("local", "true");
```
Comment (DaveCTurner) [inline qa/smoke-test-http/src/test/java/org/elasticsearch/http/ClusterStateRestCancellationIT.java:75]: :+1:

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**